### PR TITLE
Add Fleet Management Service for buses and route assignment

### LINKED ISSUE #53: Bus Service - Manage Buses and Route Assignment
**As a system developer or backend service,**  
**I want to create, retrieve, and manage buses along with their route assignments,**  
**So that the system can track which buses are operating on which routes and support future features like live tracking and ETA prediction.**

**Acceptance Criteria:**

```gherkin
Given the bus service is running,
When a new bus is created with valid fleet code and plate number,
Then the system should store the bus and return its details.

Given buses exist in the system,
When I request all buses,
Then I should receive a list of all buses with their metadata.

Given a valid route_id,
When I request buses assigned to that route,
Then only buses assigned to that route should be returned.

Given a bus exists,
When I assign it to a route,
Then the bus should be linked to that route successfully.

Given a bus exists,
When I request its details by ID,
Then the system should return the correct bus information.

Given an invalid bus ID,
When I request that bus,
Then the system should return a 404 error.

### PR DESCRIPTION
**Pull request by chamodhk**
## Summary

- Added a new Fleet Management Service for managing buses and route assignments.
- Added FastAPI endpoints to create buses, list buses, fetch a bus by ID, assign/unassign routes, and list buses by route.
- Added health check support with database connectivity validation.
- Added SQLAlchemy models, Pydantic schemas, database setup, Dockerfile, and service wiring.
- Updated Docker Compose and database initialization to include `fleet_db`.
- Added CI workflow and unit/integration tests for the fleet service.

## Testing

- Added tests for:
  - Fleet bus creation
  - Empty fleet listing
  - Health endpoint
  - Full create/assign/get-by-route flow
- CI configured to run `pytest -v tests` for `services/fleet-management-service`.

### CODE REVIEW

File: docker/docker-compose.yml [hunk 1]
```diff
@@ -10,6 +10,7 @@ services:
       - "${POSTGRES_PORT:-5432}:5432"
     volumes:
       - postgres_data:/var/lib/postgresql/data
+      - ./init:/docker-entrypoint-initdb.d
     networks:
       - ontime_network
     healthcheck:
```
Comment (JPabasara) [inline docker/docker-compose.yml:16]: docker/docker-compose.yml conflict must be resolved using main as the base. This  branch was created before the newer Kafka/MQTT/ingestion compose changes, Keep main’s broker, mqtt-broker, ingestion-service, route-service volume/command, and add only the fleet service + postgres init mount.

File: services/fleet-management-service/app/routers/fleet.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+#Get single bus
+@router.get("/{bus_id}", response_model=FleetBusResponse)
+def get_bus(bus_id: int, db: Session = Depends(get_db)):
+    bus = db.query(FleetBusORM).filter(FleetBusORM.id == bus_id).first()
+    if not bus:
+        raise HTTPException(status_code=404, detail="Bus not found")
+    return bus
+
+
+# Assign bus to route
```
Comment (JPabasara) [inline services/fleet-management-service/app/routers/fleet.py:49]: This allows assigning any route_id, even if the route does not exist. Add validation against route-service or the route database, and return 404 for invalid routes.

File: services/fleet-management-service/app/routers/fleet.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+router = APIRouter(
+    prefix="/api/v1/fleet/buses",
+    tags=["Fleet Management"]
+)
+
+
+#Create a new bus
+@router.post("", response_model=FleetBusResponse)
+def create_bus(bus: FleetBusCreate, db: Session = Depends(get_db)):
```
Comment (JPabasara) [inline services/fleet-management-service/app/routers/fleet.py:18]: Here it does not handle duplicate fleet_code / plate_number
Since the model has unique constraints, duplicates will throw a DB error/500. Catch IntegrityError, rollback, and return 409 Conflict.

File: services/fleet-management-service/tests/integration/test_api.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
+# tests/integration/test_api.py
```
Comment (JPabasara) [inline services/fleet-management-service/tests/integration/test_api.py:1]: add tests for duplicate bus creation, invalid route assignment, missing bus assign/unassign, health DB failure, and app startup/lifespan. Also tests/conftest.py:22 defaults to port 5433, while compose/CI use 5432.

File: docker/docker-compose.yml [hunk 5]
```diff
@@ -103,7 +104,25 @@ services:
     volumes:
       - ./services/route-service:/app
     environment:
-      - DATABASE_URL=postgresql://${POSTGRES_USER:-postgres}:${POSTGRES_PASSWORD:-postgres}@postgres:5432/${POSTGRES_DB:-ontime_db}
+      - DATABASE_URL=postgresql://${POSTGRES_USER:-postgres}:${POSTGRES_PASSWORD:-postgres}@postgres:5432/route_db
```
Comment (JPabasara) [inline docker/docker-compose.yml]: this line changes route-service from ${POSTGRES_DB:-ontime_db} to route_db, but existing seed scripts default to ontime_db in scripts/models/settings.py:11. This can make route-service start with an empty route DB.

**Review by JPabasara — CHANGES_REQUESTED** [review]
There is a problem that the current repo still has older shared-DB logic for buses.

If Fleet owns buses, then all bus-related logic must move to Fleet ownership.

What should change:

Keep fleet_db
Fleet Management Service owns buses.
fleet_db.buses is correct for microservice architecture.
route_id inside fleet DB should be treated as an external reference, not a SQL foreign key.
Validate route assignment through route-service
When assigning a bus:
PATCH /api/v1/fleet/buses/{bus_id}/assign-route/{route_id}
Fleet service should call route-service:

GET /api/v1/routes/{route_id}
Then:

if route exists: assign it
if route does not exist: return 404 Route not found
if route-service is down: return 503 Route service unavailable
Do not directly query route-service’s database from fleet-service.

Update old shared bus scripts

### These currently break the microservice idea:

1. **scripts/models/db_bus.py
2. scripts/seed_buses.py
3. scripts/gps_simulator.py

_They currently assume buses live in the same/shared DB as routes.**_

### They should be changed so:

bus seeding calls Fleet API or writes only to fleet_db as a fleet-owned seed script
GPS simulator gets buses from Fleet API
GPS simulator gets route geometry from Route Service API
So instead of:

select(BusORM)
it should eventually call something like:

GET /api/v1/fleet/buses
GET /api/v1/routes/{route_id}

Fix route-service /routes/{route_id}/buses
This endpoint currently belongs to route-service but returns bus data, which is fleet ownership.
Best microservice-correct option:

remove/deprecate route-service’s /api/v1/routes/{route_id}/buses
expose bus-by-route only from Fleet:
GET /api/v1/fleet/buses/route/{route_id}
Alternative if frontend needs the old endpoint:

API Gateway should aggregate it, not route-service.
Gateway calls route-service for route info and fleet-service for assigned buses.

### Compose conflict resolution
If following microservices, 
fleet-management-service:
  environment:
    - DATABASE_URL=postgresql://postgres:postgres@postgres:5432/fleet_db
But  main’s ingestion/Kafka/MQTT config while resolving the conflict.
So I recommend merge main and do necessary changes

Review comment you can send:

File: docker/docker-compose.yml (hunk 1 above)
Comment (chamodhk) [reply docker/docker-compose.yml:16]: Fleet Management Service itself does not depend on the root scripts folder. It owns buses in fleet_db, and route_id is now treated as an external reference validated through route-service.

The listed scripts are legacy local simulation/seed tooling. I agree they should be updated separately so bus seeding uses Fleet API or fleet_db only, and the GPS simulator reads buses from Fleet API and route geometry from Route Service API. I’ll keep that as a follow-up cleanup rather than mixing it into the fleet-service runtime fix.

File: services/fleet-management-service/app/routers/fleet.py (hunk 2 above)
Comment (chamodhk) [reply services/fleet-management-service/app/routers/fleet.py:49]: fixed

File: services/fleet-management-service/app/routers/fleet.py (hunk 3 above)
Comment (chamodhk) [reply services/fleet-management-service/app/routers/fleet.py:18]: fixed

File: services/fleet-management-service/tests/integration/test_api.py (hunk 4 above)
Comment (chamodhk) [reply services/fleet-management-service/tests/integration/test_api.py:1]: fixed

File: docker/docker-compose.yml (hunk 5 above)
Comment (chamodhk) [reply docker/docker-compose.yml]: I removed route-db and the route service will use ontime_db

**Review by JPabasara — APPROVED** [review]
Now it looks good I have remove some trailing whitespaces and added capacity check.
